server/client.go

```diff
@@ -3597,6 +3613,33 @@ func (c *client) setHeader(key, value string, msg []byte) []byte {
 	return bb.Bytes()
 }
 
+// Will return the value for the header denoted by key or nil if it does not exists.
+// This function ignores errors and tries to achieve speed and no additional allocations.
+func getHeader(key string, hdr []byte) []byte {
+	if len(hdr) == 0 {
+		return nil
+	}
+	index := bytes.Index(hdr, []byte(key))
+	if index < 0 {
+		return nil
+	}
+
+	var value []byte
+	hdrLen := len(hdr)
+	index += len(key) + 1
```

Comment:
In the mqtt_clustering branch I have modified getHdrVal() with the following:
```
if index >= hdrLen {
    return nil
}
if hdr[index] != ':' {
    return nil
}
index++
for hdr[index] == ' ' && index < hdrLen {
    index++
}
for index < hdrLen {
    ...
```
The reason for that is that since we don't do proper lifting, at least lets make sure that the header name matches what we are looking for. Say you look for key `foo` in the header `foobar: some_value`, I don't want this function to return `ar: somevalue` but return nil because `foo:` was not found.

Reply:
Good idea, will fix.

Reply:
ok fixed.

Reply:
Not, it is not, you did not check for `if index >= hdrLen` so then it would panic. See this: https://play.golang.org/p/98u-DLyOd-_3